### PR TITLE
add plugin to detect user sessions using loginctl

### PR DESCRIPTION
**Pull request by davide125**
Add a new plugin to expose session information on systemd hosts using loginctl. This could be extended to support consolekit and other backends as well if needed.

I'm a Facebook employee and @jaymzh will do the final merge.

### CODE REVIEW

**Comment by chef-supermarket** [comment]
Hi. I am an automated pull request bot named Curry. There are commits in this pull request whose authors are not yet authorized to contribute to Chef Software, Inc. projects or are using a non-GitHub verified email address. To become authorized to contribute, you will need to sign the Contributor License Agreement (CLA) as an individual or on behalf of your company. [You can read more on Chef's blog.](https://www.getchef.com/blog/2014/06/23/changes-to-the-contributor-license-agreement-process)

## GitHub Users Who Are Not Authorized To Contribute

The following GitHub users do not appear to have signed a CLA:

* @davide125

[Please sign the CLA here.](https://supermarket.chef.io/become-a-contributor)

**Comment by jaymzh** [comment]
:+1: 
CC: @chef/client-core

File: lib/ohai/plugins/linux/sessions.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+require "ohai/util/file_helper"
+
+include Ohai::Util::FileHelper
+
+Ohai.plugin(:Sessions) do
+  provides "sessions/by_session", "sessions/by_user"
+
+  collect_data(:linux) do
+    sessions Mash.new unless sessions
```
Comment (tas50) [inline lib/ohai/plugins/linux/sessions.rb]: The sessions mash creation should get moved down after the shellout so if we fail to shellout and gather the information we don't end up with an empty session mash sitting on the output.

File: lib/ohai/plugins/linux/sessions.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+Ohai.plugin(:Sessions) do
+  provides "sessions/by_session", "sessions/by_user"
+
+  collect_data(:linux) do
+    loginctl_path = which("loginctl")
+    if loginctl_path
+      cmd = "#{loginctl_path} --no-pager --no-legend --no-ask-password " +
+        "list-sessions"
+      loginctl = shell_out(cmd)
```
Comment (lamont-granquist) [inline lib/ohai/plugins/linux/sessions.rb]: `shell_out!` so errors throw us out of the plugin?

File: lib/ohai/plugins/linux/sessions.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+require "ohai/util/file_helper"
+
+include Ohai::Util::FileHelper
+
+Ohai.plugin(:Sessions) do
+  provides "sessions/by_session", "sessions/by_user"
+
+  collect_data(:linux) do
+    loginctl_path = which("loginctl")
+    if loginctl_path
```
Comment (lamont-granquist) [inline lib/ohai/plugins/linux/sessions.rb]: `break unless loginctl_path` would help avoid a layer of indentation

**Comment by lamont-granquist** [comment]
2 nits, but :+1:

File: lib/ohai/plugins/linux/sessions.rb (hunk 2 above)
Comment (davide125) [inline lib/ohai/plugins/linux/sessions.rb]: Yep, will do.

File: lib/ohai/plugins/linux/sessions.rb (hunk 3 above)
Comment (davide125) [inline lib/ohai/plugins/linux/sessions.rb]: This results in a test failure

```
  1) Ohai::System Linux sessions plugin should not populate sessions if loginctl is not found
     Failure/Error: break unless loginctl_path
     
     LocalJumpError:
       break from proc-closure
     # ./lib/ohai/plugins/linux/sessions.rb:28:in `block (2 levels) in <main>'
     # ./lib/ohai/dsl/plugin/versionvii.rb:88:in `instance_eval'
     # ./lib/ohai/dsl/plugin/versionvii.rb:88:in `run_plugin'
     # ./lib/ohai/dsl/plugin.rb:98:in `run'
     # ./spec/unit/plugins/linux/sessions_spec.rb:95:in `block (2 levels) in <top (required)>'

```

**Comment by lamont-granquist** [comment]
i think shell_out! doesn't actually exist in ohai.  you're probably not getting errors because you stubbed it, but should back that out.  my bad.

does the other one work if you change `break` into `return`?  i can never quite get the rules for that straight in my head...

**Comment by davide125** [comment]
Nope, it fails with return too
```
Run options:
  include {:focus=>true}
  exclude {:requires_root=>true, :ruby_19_only=>true, :ruby_18_only=>true, :windows_only=>true}

All examples were filtered out; ignoring {:focus=>true}

Ohai::System Linux sessions plugin
  should populate sessions if loginctl is found
  should not populate sessions if loginctl is not found (FAILED - 1)

Failures:

  1) Ohai::System Linux sessions plugin should not populate sessions if loginctl is not found
     Failure/Error: return unless loginctl_path
     
     LocalJumpError:
       unexpected return
     # ./lib/ohai/plugins/linux/sessions.rb:28:in `block (2 levels) in <main>'
     # ./lib/ohai/dsl/plugin/versionvii.rb:88:in `instance_eval'
     # ./lib/ohai/dsl/plugin/versionvii.rb:88:in `run_plugin'
     # ./lib/ohai/dsl/plugin.rb:98:in `run'
     # ./spec/unit/plugins/linux/sessions_spec.rb:95:in `block (2 levels) in <top (required)>'

Finished in 0.01189 seconds (files took 0.12055 seconds to load)
2 examples, 1 failure

Failed examples:

rspec ./spec/unit/plugins/linux/sessions_spec.rb:93 # Ohai::System Linux sessions plugin should not populate sessions if loginctl is not found
```

**Comment by lamont-granquist** [comment]
okay what you've got is fine then :+1:

**Comment by jaymzh** [comment]
You want "next true" instead of return, I think

**Comment by jaymzh** [comment]
(that said, I think the 'if' is fine)

**Comment by lamont-granquist** [comment]
if i have to think that hard to come up with the right code, its probably not the right code...  lets just go with `if`...  =)

**Pull request merged**